arangod/StorageEngine/VPackSortMigration.cpp

```diff
@@ -0,0 +1,294 @@
+
+#include "Indexes/Index.h"
+#include "Logger/LogMacros.h"
+#include "RestServer/DatabaseFeature.h"
+#include "RocksDBEngine/RocksDBComparator.h"
+#include "RocksDBEngine/RocksDBEngine.h"
+#include "RocksDBEngine/RocksDBIndex.h"
+#include "StorageEngine/EngineSelectorFeature.h"
+#include "StorageEngine/PhysicalCollection.h"
+#include "Utils/SingleCollectionTransaction.h"
+#include "Transaction/StandaloneContext.h"
+#include "VocBase/LogicalCollection.h"
+
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/ClusterInfo.h"
+
+#include "Network/NetworkFeature.h"
+#include "Network/Methods.h"
+
+namespace arangodb {
+
+// On dbservers, agents and single servers:
+Result analyzeVPackIndexSorting(TRI_vocbase_t& vocbase, VPackBuilder& result) {
+  // Just for the sake of completeness, restrict ourselves to the RocksDB
+  // storage engine:
+  auto& engineSelectorFeature =
+      vocbase.server().getFeature<EngineSelectorFeature>();
+  if (!engineSelectorFeature.isRocksDB()) {
+    {
+      VPackObjectBuilder guard(&result);
+      guard->add("error", VPackValue(false));
+      guard->add("errorCode", VPackValue(0));
+      guard->add(
+          "errorMessage",
+          VPackValue("VPack sorting migration is unnecessary for storage "
+                     "engines other than RocksDB"));
+    }
+    return {};
+  }
+
+  auto& engine = engineSelectorFeature.engine<RocksDBEngine>();
+  auto* db = engine.db();
+
+  using IndexType = arangodb::Index::IndexType;
+  DatabaseFeature& databaseFeature =
+      vocbase.server().getFeature<DatabaseFeature>();
+  auto newComparator = RocksDBVPackComparator<
+      arangodb::basics::VelocyPackHelper::SortingMethod::Correct>();
+
+  // Collect problems here:
+  result.clear();
+  {
+    VPackObjectBuilder guardO(&result);
+    bool problemFound = false;
+    guardO->add(VPackValue("affected"));  // just write key
+    {
+      VPackArrayBuilder guardA(&result);
+
+      for (auto const& name : databaseFeature.getDatabaseNames()) {
+        auto database = databaseFeature.useDatabase(name);
+        if (database == nullptr) {
+          continue;
+        }
+        LOG_TOPIC("76521", DEBUG, Logger::ENGINES)
+            << "Checking VPackSortMigration for database " << name;
+        database->processCollections([&](LogicalCollection* collection) {
+          LOG_TOPIC("42537", DEBUG, Logger::ENGINES)
+              << "Checking VPackSortMigration for collection "
+              << collection->name();
+          for (auto& index : collection->getPhysical()->getReadyIndexes()) {
+            LOG_TOPIC("42538", DEBUG, Logger::ENGINES)
+                << "Checking VPackSortMigration for index with ID "
+                << index->id().id() << " and name " << index->name();
+            if (auto type = index->type();
+                type == IndexType::TRI_IDX_TYPE_PERSISTENT_INDEX ||
+                type == IndexType::TRI_IDX_TYPE_HASH_INDEX ||
+                type == IndexType::TRI_IDX_TYPE_SKIPLIST_INDEX ||
```

Comment:
we may want to include TTL indexes here too. TTL indexes are based on persistent indexes and also use the VPackComparator.

Reply:
Good point, will look into this.